nixos-rebuild: prevent build result from being garbage-collected in switch

dpausp opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The system built in `nixos-rebuild switch` should be protected from garbage-collection in all cases.
When not using the flake, create a result link in the temp dir. The flake case already does this.

 Fixes #118804

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - nixosTests.installer 
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ctheune commented:
For good measure: I've reviewed this change internally with @dpausp internally (at the Flying Circus) but neither of us has commit rights until now and I'm also not too keen to perform a "second pair of eyes" review while both us work at the same organisation. :)

dpausp commented:
I can fix the conflicts after #123720 has been merged.

jonringer commented:
other was merged

elikoga committed:
Prevent build result from being garbage-collected in nixos-rebuild switch

When not using the flake, create a result link in the temp dir
to protect the built system. The flake case already does this.

Co-authored-by: Tobias Stenzel <ts@flyingcircus.io>

chvp reviewed (commented):
Did you test the flake case? It would seem weird to me that it still works without defining `pathToConfig` in that case? How does this interact with remote building?
